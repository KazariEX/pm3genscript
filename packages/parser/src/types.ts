export interface ParseError {
    message: string;
    offset: number;
}

export interface Token {
    type: TokenType;
    value: string;
    offset: number;
}

export interface TokenTypeMap {
    hash: "hash";
    at: "at";
    equal: "equal";
    identifier: "identifier";
    number: "number";
    string: "string";
}

export type TokenType = TokenTypeMap[keyof TokenTypeMap];

export interface Node {
    type: string;
    offset: number;
}

export interface Root extends Node {
    type: "root";
    children: Node[];
}

export interface Macro extends Node {
    type: "macro";
    name: Identifier;
    arguments: Argument[];
}

export interface Command extends Node {
    type: "command";
    name: Identifier;
    arguments: Argument[];
}

export interface Dynamic extends Node {
    type: "dynamic";
    name: Identifier;
}

export interface Identifier extends Node {
    type: "identifier";
    value: string;
}

export interface NumberLiteral extends Node {
    type: "number";
    value: string;
}

export interface StringLiternal extends Node {
    type: "string";
    value: string;
}

export type Argument = Dynamic | Identifier | NumberLiteral | StringLiternal;