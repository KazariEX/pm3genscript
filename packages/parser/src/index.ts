export * from "./parse";
export * from "./tokenize";
export * from "./types";