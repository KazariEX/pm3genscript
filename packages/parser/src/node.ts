export abstract class Node {
    constructor(
        public type: string,
        public offset: number
    ) {}

    getText(root: Root) {
        return root.text.slice(this.offset, this.getEnd());
    }

    abstract getEnd(): number;

    getLength() {
        return this.getEnd() - this.offset;
    }
}

export class Root extends Node {
    children: Node[] = [];

    constructor(
        public text: string
    ) {
        super("root", 0);
    }

    override getText() {
        return this.text;
    }

    getEnd() {
        return this.text.length;
    }
}

export abstract class Parent extends Node {
    arguments: Argument[] = [];

    constructor(
        public override type: string,
        public override offset: number,
        public name: Identifier
    ) {
        super(type, offset);
    }

    getEnd() {
        return this.arguments.length ? this.arguments.at(-1)!.getEnd() : this.name.getEnd();
    }
}

export class Macro extends Parent {
    constructor(
        public override offset: number,
        public override name: Identifier
    ) {
        super("macro", offset, name);
    }
}

export class Command extends Parent {
    constructor(
        public override name: Identifier
    ) {
        super("command", name.offset, name);
    }
}

export class Dynamic extends Node {
    constructor(
        public override offset: number,
        public name: Identifier
    ) {
        super("dynamic", offset);
    }

    getEnd() {
        return this.name.getEnd();
    }
}

export abstract class Literal extends Node {
    constructor(
        public override type: string,
        public override offset: number,
        public value: string
    ) {
        super(type, offset);
    }

    override getText() {
        return this.value;
    }

    getEnd() {
        return this.offset + this.value.length;
    }

    override getLength() {
        return this.value.length;
    }
}

export class Identifier extends Literal {
    constructor(
        public override offset: number,
        public override value: string
    ) {
        super("identifier", offset, value);
    }
}

export class NumberLiteral extends Literal {
    constructor(
        public override offset: number,
        public override value: string
    ) {
        super("number", offset, value);
    }
}

export class StringLiternal extends Literal {
    constructor(
        public override offset: number,
        public override value: string
    ) {
        super("string", offset, value);
    }
}

export type Argument = Dynamic | Identifier | NumberLiteral | StringLiternal;

export class Block extends Node {
    children: Command[] = [];

    constructor(
        public label: Macro
    ) {
        super("block", label.offset);
    }

    getEnd() {
        return this.children.length ? this.children.at(-1)!.getEnd() : this.label.getEnd();
    }
}