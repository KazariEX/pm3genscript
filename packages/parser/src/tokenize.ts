import type { ParseError, Token, TokenType } from "./types";

export function tokenize(text: string) {
    const tokens: Token[] = [];
    const errors: ParseError[] = [];
    const textLength = text.length;
    let idx = 0;

    while (idx < textLength) {
        const char = text[idx];
        switch (char) {
            case "#": {
                add("hash", char, idx);
                idx++;
                break;
            }
            case "@": {
                add("at", char, idx);
                idx++;
                break;
            }
            case "=": {
                add("equal", char, idx);
                idx++;
                break;
            }
            default: {
                const offset = idx;
                let value = "";
                let next = char;
                if (isWhitespace(char)) {
                    do {
                        next = text[++idx];
                    } while (isWhitespace(next) && idx < textLength);
                    break;
                }
                else if (isAlpha(char)) {
                    do {
                        value += next;
                        next = text[++idx];
                    } while ((isAlphaNumeric(next) || next === "_") && idx < textLength);
                    add("identifier", value, offset);
                    break;
                }
                else if (isDigit(char)) {
                    let isHex = false;
                    do {
                        value += next;
                        next = text[++idx];
                        if (value.length === 1 && next === "x") {
                            isHex = true;
                            value += next;
                            next = text[++idx];
                        }
                    } while ((isHex ? isHexNumeric(next) : isDigit(next)) && idx < textLength);
                    add("number", value, offset);
                    break;
                }
                else if (char === "\"" || char === "'") {
                    const quote = char;
                    do {
                        value += next;
                        next = text[++idx];
                    } while (next !== quote && idx < textLength);
                    value += next;
                    idx++;
                    add("string", value, offset);
                    break;
                }
                else {
                    errors.push({
                        message: `Unexpected character "${char}".`,
                        offset: idx
                    });
                }
            }
        }
    }

    return {
        tokens,
        errors
    };

    function add(type: TokenType, value: string, offset: number) {
        tokens.push({
            type,
            value,
            offset
        });
    }
}

function isWhitespace(char: string) {
    return char === " " || char === "\n" || char === "\r" || char === "\t";
}

function isAlpha(char: string) {
    const code = char.charCodeAt(0);
    return code >= 97 && code <= 122 || code >= 65 && code <= 90;
}

function isDigit(char: string) {
    const code = char.charCodeAt(0);
    return code >= 48 && code <= 57;
}

function isAlphaNumeric(char: string) {
    const code = char.charCodeAt(0);
    return code >= 97 && code <= 122
        || code >= 65 && code <= 90
        || code >= 48 && code <= 57;
}

function isHexNumeric(char: string) {
    const code = char.charCodeAt(0);
    return code >= 97 && code <= 102
        || code >= 65 && code <= 70
        || code >= 48 && code <= 57;
}