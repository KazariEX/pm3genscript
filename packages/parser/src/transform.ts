import { commands, macros } from "@pm3genscript/shared";
import { Block, Command, Dynamic, Macro, type Root } from "./node";
import type { AST, Diagnostic } from "./types";

export function transform(root: Root) {
    const ast: AST = {
        dynamic: {
            defines: [],
            references: []
        },
        children: []
    };
    const diagnostics: Diagnostic[] = [];
    const childLength = root.children.length;
    let current = 0;
    let currentBlock: Block | null = null;

    while (current < childLength) {
        const child = root.children[current];
        if (child instanceof Macro) {
            transformMacro(child);
        }
        else if (child instanceof Command) {
            transformCommand(child);
        }
        else {
            diagnostics.push({
                message: `Expected "macro" or "command" node at the root, got "${child.type}"`,
                offset: child.offset,
                length: child.getLength()
            });
        }
        current++;
    }

    return {
        ast,
        diagnostics
    };

    function transformMacro(macro: Macro) {
        const name = macro.name.value;
        switch (name) {
            case "org": {
                const block = new Block(macro);
                currentBlock = block;
                ast.children.push(block);

                const arg0 = macro.arguments[0];
                if (arg0 instanceof Dynamic) {
                    ast.dynamic.defines.push(arg0);
                }
                break;
            }
            default: {
                currentBlock = null;
                ast.children.push(macro);
                break;
            }
        }

        const template = getMacroTemplate(name);
        if (template) {
            const args = macro.arguments;
            const templateArgs = template.arguments;
            if (args.length !== templateArgs.length) {
                diagnostics.push({
                    message: `Expected ${templateArgs.length} arguments, got ${args.length}.`,
                    offset: macro.offset,
                    length: macro.getLength()
                });
            }
        }
        else {
            diagnostics.push({
                message: `Unknown macro "${name}".`,
                offset: macro.name.offset,
                length: macro.name.getLength()
            });
        }
    }

    function transformCommand(command: Command) {
        const name = command.name.value;
        const template = getCommandTemplate(name)!;
        const argLength = command.arguments.length;
        const templateArgLength = template.arguments?.length ?? 0;
        if (argLength !== templateArgLength) {
            diagnostics.push({
                message: `Expected ${templateArgLength} arguments, got ${argLength}.`,
                offset: command.offset,
                length: command.getLength()
            });
        }

        for (const arg of command.arguments) {
            if (arg instanceof Dynamic) {
                ast.dynamic.references.push(arg);
            }
        }

        if (currentBlock) {
            currentBlock.children.push(command);
        }
        else {
            ast.children.push(command);
            diagnostics.push({
                message: `Command "${name}" is not inside a block.`,
                offset: command.offset,
                length: command.getLength()
            });
        }
    }
}

function getMacroTemplate(name: string) {
    let template = macros[name];
    if (template?.redirect) {
        template = macros[template.redirect];
    }
    return template;
}

function getCommandTemplate(name: string) {
    let template = commands[name];
    if (template?.redirect) {
        template = commands[template.redirect];
    }
    return template;
}