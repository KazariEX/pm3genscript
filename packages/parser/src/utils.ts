import { Block, Command, Dynamic, Macro, type Node, Parent, Root } from "./node";

export function forEachNode(
    node: Node,
    visit: (node: Node, parent: Node | undefined, parents: Node[]) => void
) {
    const parents: Node[] = [];
    traverse(node);

    function traverse(node: Node) {
        visit(node, parents[parents.length - 1], parents);

        parents.push(node);
        if (node instanceof Root) {
            for (const child of node.children) {
                traverse(child);
            }
        }
        else if (node instanceof Block) {
            traverse(node.label);
            for (const child of node.children) {
                traverse(child);
            }
        }
        else if (node instanceof Parent) {
            traverse(node.name);
            for (const arg of node.arguments) {
                traverse(arg);
            }
        }
        else if (node instanceof Dynamic) {
            traverse(node.name);
        }
        parents.pop();
    }
}

export function* forEachStatement(children: Node[]): Generator<Macro | Command> {
    for (const child of children) {
        if (child instanceof Block) {
            yield child.label;
            yield* child.children;
        }
        else if (child instanceof Macro || child instanceof Command) {
            yield child;
        }
    }
}