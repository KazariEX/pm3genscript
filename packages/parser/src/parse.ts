import { commands } from "@pm3genscript/shared";
import { tokenize } from "./tokenize";
import type { Argument, Command, Dynamic, Identifier, Macro, NumberLiteral, ParseError, Root, StringLiternal, Token } from "./types";

export function parse(text: string) {
    const tokenized = tokenize(text);
    const walked = walkTokens(tokenized.tokens);

    return {
        root: walked.root,
        errors: [
            ...tokenized.errors,
            ...walked.errors
        ]
    };
}

export function walkTokens(tokens: Token[]) {
    const root: Root = {
        type: "root",
        offset: 0,
        children: []
    };
    const errors: ParseError[] = [];
    const tokenLength = tokens.length;
    let current = 0;

    const parents: (Macro | Command)[] = [];

    while (current < tokenLength) {
        const token = tokens[current];

        switch (token.type) {
            case "equal":
            case "hash": {
                const macro: Macro = {
                    type: "macro",
                    offset: token.offset,
                    name: null!,
                    arguments: []
                };
                root.children.push(macro);
                parents.unshift(macro);

                const next = advance();
                if (token.type === "equal") {
                    macro.name = createIdentifier(token);
                }
                else if (next?.type === "identifier") {
                    macro.name = createIdentifier(next);
                    current++;
                }
                else {
                    macro.name = {
                        type: "identifier",
                        offset: token.offset + token.value.length,
                        value: ""
                    };
                    errors.push({
                        message: `Expected identifier after "#", got "${next.type}".`,
                        offset: next.offset
                    });
                }
                break;
            }
            case "at": {
                const dynamic: Dynamic = {
                    type: "dynamic",
                    offset: token.offset,
                    name: null!
                };
                attach(dynamic);

                const next = advance();
                if (next?.type === "identifier" || next?.type === "number") {
                    dynamic.name = createIdentifier(next);
                    current++;
                }
                else {
                    dynamic.name = {
                        type: "identifier",
                        offset: token.offset + token.value.length,
                        value: ""
                    };
                    errors.push({
                        message: `Expected identifier or number after "@", got "${next.type}".`,
                        offset: next.offset
                    });
                }
                break;
            }
            case "identifier": {
                const identifier = createIdentifier(token);
                if (identifier.value in commands) {
                    const command: Command = {
                        type: "command",
                        offset: identifier.offset,
                        name: identifier,
                        arguments: []
                    };
                    parents.unshift(command);
                    root.children.push(command);
                }
                else {
                    attach(identifier);
                }
                current++;
                break;
            }
            case "number": {
                const number = createNumberLiteral(token);
                attach(number);
                current++;
                break;
            }
            case "string": {
                const string = createStringLiteral(token);
                attach(string);
                current++;
                break;
            }
        }
    }

    return {
        root,
        errors
    };

    function advance() {
        return tokens[++current];
    }

    function attach(node: Argument) {
        const container = parents.length ? parents[0].arguments : root.children;
        container.push(node);
    }
}

function createIdentifier(token: Token): Identifier {
    return {
        type: "identifier",
        offset: token.offset,
        value: token.value
    };
}

function createNumberLiteral(token: Token): NumberLiteral {
    return {
        type: "number",
        offset: token.offset,
        value: token.value
    };
}

function createStringLiteral(token: Token): StringLiternal {
    return {
        type: "string",
        offset: token.offset,
        value: token.value
    };
}