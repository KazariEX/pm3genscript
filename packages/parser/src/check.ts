import { type ArgumentType, argumentTypes } from "@pm3genscript/shared";
import { isBlock, isCommand, isDynamic, isIdentifier, isMacro, isNumberLiteral, isStringLiteral, isSymbol } from "./utils/is";
import { joinWords, joinWordsOr, toArray } from "./utils/shared";
import { forEachNode, forEachStatement } from "./utils/traverse";
import type { Argument, Command, Dynamic, Macro, Parent, Root, Symbol } from "./node";
import type { Diagnostic } from "./types";

export function check(ast: Root) {
    const diagnostics: Diagnostic[] = [];
    const resolvedValues = new Map<Argument, any>();
    const dynamics = new Map<string, {
        macro: Macro;
        definition: Dynamic;
        references: Dynamic[];
    }>();
    const symbols = new Map<string, {
        macro: Macro;
        definition: Symbol;
        references: Symbol[];
        type: ArgumentType;
        value: any;
    }>();

    const references = {
        dynamics: [] as Dynamic[],
        symbols: [] as Symbol[]
    };

    for (const node of ast.children) {
        if (isMacro(node) || isBlock(node)) {
            continue;
        }
        else if (isCommand(node)) {
            diagnostics.push({
                message: `Command "${node.name.value}" is not inside a block.`,
                offset: node.offset,
                length: node.getLength()
            });
        }
        else {
            diagnostics.push({
                message: `Expected "macro" or "command" node at the root, got "${node.type}"`,
                offset: node.offset,
                length: node.getLength()
            });
        }
    }

    forEachNode(ast, (node, parent) => {
        if (isDynamic(node)) {
            if (
                parent && isMacro(parent) &&
                parent.canonicalName === "org" &&
                parent.arguments[0] === node
            ) {
                const name = node.name.value;
                if (!dynamics.has(name)) {
                    dynamics.set(name, {
                        macro: parent,
                        definition: node,
                        references: []
                    });
                }
                else {
                    diagnostics.push({
                        message: `Dynamic offset "@${name}" is already defined.`,
                        offset: node.offset,
                        length: node.getLength()
                    });
                }
            }
            else {
                references.dynamics.push(node);
            }
        }
        else if (isSymbol(node)) {
            if (
                parent && isMacro(parent) &&
                parent.canonicalName === "define" &&
                parent.arguments[0] === node &&
                parent.arguments.length >= 2
            ) {
                const [, arg1] = parent.arguments;
                const name = node.value;
                const [type, value] = resolveRuntimeTypeAndValue(arg1);
                if (symbols.has(name)) {
                    symbols.get(name)!.definition = node;
                }
                else {
                    symbols.set(name, {
                        macro: parent,
                        definition: node,
                        references: [],
                        type,
                        value
                    });
                }
            }
            else {
                references.symbols.push(node);
            }
        }
    });

    for (const node of references.dynamics) {
        const name = node.name.value;
        if (dynamics.has(name)) {
            dynamics.get(name)!.references.push(node);
        }
        else if (name) {
            diagnostics.push({
                message: `Dynamic offset "@${name}" is not defined.`,
                offset: node.offset,
                length: node.getLength()
            });
        }
    }

    for (const node of references.symbols) {
        const name = node.value;
        if (symbols.has(name)) {
            symbols.get(name)!.references.push(node);
        }
        else {
            diagnostics.push({
                message: `Symbol "${name}" is not defined.`,
                offset: node.offset,
                length: node.getLength()
            });
        }
    }

    for (const node of forEachStatement(ast.children)) {
        isMacro(node) ? checkMacro(node) : checkCommand(node);
    }

    return {
        diagnostics,
        resolvedValues,
        dynamics,
        symbols
    };

    function checkMacro(macro: Macro) {
        const { template } = macro;
        if (!template) {
            if (macro.name.value) {
                diagnostics.push({
                    message: `Unknown macro "${macro.name.value}".`,
                    offset: macro.name.offset,
                    length: macro.name.getLength()
                });
            }
            return;
        }

        if (macro.canonicalName === "raw") {
            checkRaw(macro);
            return;
        }

        const { arguments: args } = macro;
        const { arguments: templateArgs = [] } = template;
        const length = Math.max(args.length, templateArgs.length);

        for (let i = 0; i < length; i++) {
            const arg = args[i];
            const templateArg = templateArgs[i];
            if (!arg || !templateArg) {
                mismatchArgumentCount(macro, templateArgs.length);
                break;
            }

            const allowedTypes = getAllowedTypes(templateArg.type ?? []);
            const canSymbol = (templateArg.can?.symbol ?? true) && !allowedTypes.has("symbol");
            const [type, value] = resolveRuntimeTypeAndValue(arg, canSymbol);

            if (templateArg.enum) {
                if (!templateArg.enum.includes(value)) {
                    mismatchArgumentEnum(arg, templateArg.enum);
                }
            }
            else if (!allowedTypes.has(type)) {
                mismatchArgumentType(arg, type, templateArg.type ?? []);
            }
        }
    }

    function checkRaw(macro: Macro) {
        const { arguments: args } = macro;
        const [templateArg0, templateArg1] = macro.template.arguments!;

        let prev: string | undefined;
        for (let i = 0; i < args.length; i++) {
            const arg = args[i];
            const [type, value] = resolveRuntimeTypeAndValue(arg);
            if (type === "identifier") {
                if (!templateArg0.enum!.includes(value)) {
                    mismatchArgumentEnum(arg, templateArg0.enum!);
                }
                prev = normalizeRawType(value);
                continue;
            }
            else {
                const originalType = prev && argumentTypes.has(prev) ? prev as ArgumentType : templateArg1.type!;
                const allowedTypes = getAllowedTypes(originalType);
                if (!allowedTypes.has(type)) {
                    mismatchArgumentType(arg, type, originalType);
                }
            }
            prev = void 0;
        }
    }

    function checkCommand(command: Command) {
        const { arguments: args, template } = command;
        const { arguments: templateArgs = [] } = template;
        const length = Math.max(args.length, templateArgs.length);
        const values: any[] = [];

        let idx = 0;
        let templateIdx = 0;
        for (let i = 0; i < length; i++) {
            const templateArg = templateArgs[templateIdx];
            if (templateArg && (!templateArg.when || templateArg.when(values))) {
                templateIdx++;
            }

            const arg = args[idx];
            if (arg) {
                const allowedTypes = getAllowedTypes(templateArg?.type ?? []);
                const [type, value] = resolveRuntimeTypeAndValue(arg, !allowedTypes.has("symbol"));
                values.push(value);
                idx++;

                if (templateArg && !allowedTypes.has(type)) {
                    mismatchArgumentType(arg, type, templateArg.type);
                }
            }
        }
        if (args.length !== templateIdx) {
            mismatchArgumentCount(command, templateIdx);
        }
    }

    function mismatchArgumentCount(parent: Parent, expectedLength: number) {
        const { arguments: args } = parent;
        const isMoreThan = args.length > expectedLength;
        const start = (isMoreThan ? args[expectedLength] : parent).offset;
        const end = (isMoreThan ? args.at(-1)! : parent).getEnd();
        diagnostics.push({
            message: `Expected ${expectedLength} argument(s), got ${args.length}.`,
            offset: start,
            length: end - start
        });
    }

    function mismatchArgumentEnum(arg: Argument, expectedValues: string[]) {
        diagnostics.push({
            message: `Value is not accepted. Valid values: ${joinWords(expectedValues)}.`,
            offset: arg.offset,
            length: arg.getLength()
        });
    }

    function mismatchArgumentType(arg: Argument, type: ArgumentType, expectedType: ArgumentType | ArgumentType[]) {
        const types = toArray(expectedType);
        diagnostics.push({
            message: `Expected argument type ${joinWordsOr(types)}, got "${type}".`,
            offset: arg.offset,
            length: arg.getLength()
        });
    }

    function resolveRuntimeTypeAndValue(arg: Argument, canSymbol = true): [ArgumentType, any] {
        if (isIdentifier(arg)) {
            const name = arg.value;
            resolvedValues.set(arg, name);
            return ["identifier", name];
        }
        if (isSymbol(arg)) {
            const name = arg.value;
            if (canSymbol) {
                const symbol = symbols.get(name);
                if (symbol) {
                    return [symbol.type, symbol.value];
                }
            }
            return ["symbol", name];
        }
        if (isNumberLiteral(arg)) {
            const value = arg.value.startsWith("0x")
                ? Number.parseInt(arg.value.slice(2), 16)
                : Number.parseInt(arg.value);
            resolvedValues.set(arg, value);

            const type = value >= 0 && value <= 0xFF ? "byte"
                : value >= 0 && value <= 0xFFFF ? "word"
                : value >= 0 && value <= 0xFFFFFFFF ? "dword"
                : "number";
            return [type, value];
        }
        if (isStringLiteral(arg)) {
            const value = arg.value;
            resolvedValues.set(arg, value);
            return ["string", value];
        }
        // HACK: 放开头会导致不可预测的类型收缩行为
        if (isDynamic(arg)) {
            const name = arg.name.value;
            resolvedValues.set(arg, name);
            return ["pointer", name];
        }
        else {
            throw arg satisfies never;
        }
    }
}

function getAllowedTypes(type: ArgumentType | ArgumentType[]) {
    const types = [...toArray(type)];
    if (types.includes("number") || types.includes("dword") || types.includes("pointer")) {
        types.push("byte", "word", "dword", "pointer");
    }
    else if (types.includes("word")) {
        types.push("byte");
    }
    return new Set(types);
}

function normalizeRawType(type: string) {
    switch (type) {
        case "b":
        case "char":
            return "byte";
        case "i":
        case "int":
        case "integer":
            return "word";
        case "l":
        case "long":
            return "dword";
        case "p":
        case "ptr":
            return "pointer";
        default:
            return type;
    }
}