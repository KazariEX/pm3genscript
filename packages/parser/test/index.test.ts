import { describe, expect, it } from "vitest";
import { walkTokens } from "../src/parse";
import { tokenize } from "../src/tokenize";

describe("basic", () => {
    const text = `
#dynamic 0xA00000

#org @1
msgbox @2 MSG_FACE
end

#org @2
= "你自己魔怔吧，我要去二次元了。"
`.trim();

    const { tokens, errors: tokenizedErrors } = tokenize(text);
    const { root, errors: walkedErrors } = walkTokens(tokens);

    it("tokenize", () => {
        expect(tokenizedErrors.length).toEqual(0);
        expect(tokens).toMatchSnapshot();
    });

    it("parse", () => {
        expect(walkedErrors.length).toEqual(0);
        expect(root).toMatchSnapshot();
    });
});