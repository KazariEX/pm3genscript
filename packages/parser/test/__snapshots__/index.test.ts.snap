// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic > parse 1`] = `
{
  "children": [
    {
      "arguments": [
        {
          "offset": 9,
          "type": "number",
          "value": "0xA00000",
        },
      ],
      "name": {
        "offset": 1,
        "type": "identifier",
        "value": "dynamic",
      },
      "offset": 0,
      "type": "macro",
    },
    {
      "arguments": [
        {
          "name": {
            "offset": 25,
            "type": "identifier",
            "value": "1",
          },
          "offset": 24,
          "type": "dynamic",
        },
      ],
      "name": {
        "offset": 20,
        "type": "identifier",
        "value": "org",
      },
      "offset": 19,
      "type": "macro",
    },
    {
      "arguments": [
        {
          "name": {
            "offset": 35,
            "type": "identifier",
            "value": "2",
          },
          "offset": 34,
          "type": "dynamic",
        },
        {
          "offset": 37,
          "type": "identifier",
          "value": "MSG_FACE",
        },
      ],
      "name": {
        "offset": 27,
        "type": "identifier",
        "value": "msgbox",
      },
      "offset": 27,
      "type": "command",
    },
    {
      "arguments": [],
      "name": {
        "offset": 46,
        "type": "identifier",
        "value": "end",
      },
      "offset": 46,
      "type": "command",
    },
    {
      "arguments": [
        {
          "name": {
            "offset": 57,
            "type": "identifier",
            "value": "2",
          },
          "offset": 56,
          "type": "dynamic",
        },
      ],
      "name": {
        "offset": 52,
        "type": "identifier",
        "value": "org",
      },
      "offset": 51,
      "type": "macro",
    },
    {
      "arguments": [
        {
          "offset": 61,
          "type": "string",
          "value": ""你自己魔怔吧，我要去二次元了。"",
        },
      ],
      "name": {
        "offset": 59,
        "type": "identifier",
        "value": "=",
      },
      "offset": 59,
      "type": "macro",
    },
  ],
  "offset": 0,
  "type": "root",
}
`;

exports[`basic > tokenize 1`] = `
[
  {
    "offset": 0,
    "type": "hash",
    "value": "#",
  },
  {
    "offset": 1,
    "type": "identifier",
    "value": "dynamic",
  },
  {
    "offset": 9,
    "type": "number",
    "value": "0xA00000",
  },
  {
    "offset": 19,
    "type": "hash",
    "value": "#",
  },
  {
    "offset": 20,
    "type": "identifier",
    "value": "org",
  },
  {
    "offset": 24,
    "type": "at",
    "value": "@",
  },
  {
    "offset": 25,
    "type": "number",
    "value": "1",
  },
  {
    "offset": 27,
    "type": "identifier",
    "value": "msgbox",
  },
  {
    "offset": 34,
    "type": "at",
    "value": "@",
  },
  {
    "offset": 35,
    "type": "number",
    "value": "2",
  },
  {
    "offset": 37,
    "type": "identifier",
    "value": "MSG_FACE",
  },
  {
    "offset": 46,
    "type": "identifier",
    "value": "end",
  },
  {
    "offset": 51,
    "type": "hash",
    "value": "#",
  },
  {
    "offset": 52,
    "type": "identifier",
    "value": "org",
  },
  {
    "offset": 56,
    "type": "at",
    "value": "@",
  },
  {
    "offset": 57,
    "type": "number",
    "value": "2",
  },
  {
    "offset": 59,
    "type": "equal",
    "value": "=",
  },
  {
    "offset": 61,
    "type": "string",
    "value": ""你自己魔怔吧，我要去二次元了。"",
  },
]
`;
